src/lib/apiTypesV2/Entity.cpp

```diff
@@ -104,72 +225,146 @@ std::string Entity::render
+  filterAttributes(attrsFilter, uriParamOptions[DATE_CREATED], uriParamOptions[DATE_MODIFIED]);
+
+  std::string out;
+  switch (renderFormat)
   {
-    ContextAttribute* caP = new ContextAttribute(DATE_MODIFIED, DATE_TYPE, modDate);
-    attributeVector.push_back(caP);
+  case NGSI_V2_VALUES:
+    out = toJsonValues();
+    break;
+  case NGSI_V2_UNIQUE_VALUES:
+    out = toJsonUniqueValues();
+    break;
+  case NGSI_V2_KEYVALUES:
+    out = toJsonKeyvalues();
+    break;
+  default:  // NGSI_V2_NORMALIZED
+    out = toJsonNormalized(metadataFilter);
+    break;
   }
 
-  if ((renderFormat == NGSI_V2_VALUES) || (renderFormat == NGSI_V2_UNIQUE_VALUES))
+  return out;
+}
+
+
+
+/* ****************************************************************************
+*
+* Entity::toJsonValues -
+*/
+std::string Entity::toJsonValues(void)
+{
+  std::string out = "[";
+
+  for (unsigned int ix = 0; ix < attributeVector.size(); ix++)
   {
-    out = "[";
-    if (attributeVector.size() != 0)
+    ContextAttribute* caP = attributeVector[ix];
+    out += caP->toJsonValue();
+
+    if (ix != attributeVector.size() - 1)
     {
-      out += attributeVector.toJson(renderFormat, attrsFilter, metadataFilter, false);
+      out += ",";
     }
-    out += "]";
   }
-  else
-  {
-    out = "{";
 
-    if (renderId)
-    {
-      out += JSON_VALUE("id", id);
-      out += ",";
+  out += "]";
 
-      /* This is needed for entities coming from NGSIv1 (which allows empty or missing types) */
-      out += JSON_STR("type") + ":" + ((type != "")? JSON_STR(type) : JSON_STR(DEFAULT_ENTITY_TYPE));
-    }
+  return out;
+}
+
+
+
+/* ****************************************************************************
+*
+* Entity::toJsonUniqueValues -
+*/
+std::string Entity::toJsonUniqueValues(void)
+{
+  std::string out = "[";
+
+  std::map<std::string, bool>  uniqueMap;
+
+  for (unsigned int ix = 0; ix < attributeVector.size(); ix++)
+  {
+    ContextAttribute* caP = attributeVector[ix];
 
-    std::string attrsOut;
-    if (attributeVector.size() != 0)
+    std::string value = caP->toJsonValue();
+
+    if (uniqueMap[value] == true)
     {
-      attrsOut += attributeVector.toJson(renderFormat, attrsFilter, metadataFilter, false);
+      // Already rendered. Skip.
+      continue;
     }
-
-    //
-    // Note that just attributeVector.size() != 0 (used in previous versions) cannot be used
-    // as ciP->uriParam["attrs"] filter could remove all the attributes
-    //
-    if (attrsOut != "")
+    else
     {
-      if (renderId)
-      {
-        out +=  "," + attrsOut;
-      }
-      else
-      {
-        out += attrsOut;
-      }
+      out += value;
+      uniqueMap[value] = true;
     }
 
-    out += "}";
+    out += ",";
   }
 
-  if (comma)
+  // The substrig trick removes the final ",". It is not very smart, but it saves
+  // a second pass on the vector, once the "unicity" has been calculated in the hashmap
+  return out.substr(0, out.length() - 1 ) + "]";
```

Comment:
A bit cryptic. Took me a few seconds to realize what was done here.
a comment with "repace the last comma with an end-bracket" would help

Reply:
Fixed in 295296b